ledger/src/state/ledger.rs

```diff
@@ -558,8 +558,31 @@ impl<N: Network> LedgerState<N> {
         let amount = Block::<N>::block_reward(block_height);
         let coinbase_transaction = Transaction::<N>::new_coinbase(recipient, amount, rng)?;
 
+        // Filter the transactions to ensure they are new, and append the coinbase transaction.
+        // TODO (howardwu): Improve the performance and design of this.
+        let mut transactions: Vec<Transaction<N>> = transactions
+            .iter()
+            .filter(|transaction| {
+                for serial_number in transaction.serial_numbers() {
+                    if let Ok(true) = self.contains_serial_number(serial_number) {
+                        return false;
+                    }
+                }
+
+                for commitment in transaction.commitments() {
+                    if let Ok(true) = self.contains_commitment(commitment) {
+                        return false;
+                    }
+                }
+
+                true
+            })
+            .cloned()
+            .collect();
```

Comment:
We should craft the coinbase transaction here and include the transaction fees in the reward.

Reply:
Will add in a followup PR.